proof/invariant-abstract/DetSchedSchedule_AI.thy

```diff
@@ -5654,10 +5680,92 @@ lemma update_time_stamp_valid_sched_misc[wp]:
+        (\<lambda>s. cur_time s = prev_time))" in hoare_weaken_pre[rotated])
+   apply clarsimp
+  apply (rule hoare_strengthen_post)
+   apply (rule dmo_getCurrentTime_sp)
+   apply wpsimp
+  apply (clarsimp simp: vs_all_heap_simps cfg_bounded_release_time_def word_le_nat_alt)
+  done
+
+lemma update_time_stamp_active_sc_valid_refills[wp]:
+  "\<lbrace>active_sc_valid_refills and valid_machine_time\<rbrace>
+   update_time_stamp
+   \<lbrace>\<lambda>_. active_sc_valid_refills :: 'state_ext state \<Rightarrow> _\<rbrace>"
+  apply (clarsimp simp: active_sc_valid_refills_def)
+  apply (wpsimp wp: hoare_vcg_all_lift hoare_vcg_imp_lift')
+  done
+
+lemma preemption_point_active_sc_valid_refills[wp]:
+  "\<lbrace>active_sc_valid_refills and valid_machine_time\<rbrace>
+   preemption_point
+   \<lbrace>\<lambda>_. active_sc_valid_refills  :: 'state_ext state \<Rightarrow> _\<rbrace>"
+  apply (wpsimp simp: preemption_point_def
+                  wp: OR_choiceE_weak_wp hoare_drop_imps)
+  done
+
+lemma invoke_domain_no_exception[wp]:
+  "\<lbrace>\<top>\<rbrace> invoke_domain thread domain -, \<lbrace>\<lambda>rv. P\<rbrace>"
+  by (wpsimp simp: invoke_domain_def)
+
+crunches update_time_stamp
+  for pred_tcb_at[wp]: "\<lambda>s. Q (pred_tcb_at p P t s)"
+
+lemma update_time_stamp_budget_ready[wp]:
+ "\<lbrace>budget_ready t and valid_machine_time :: 'state_ext state \<Rightarrow> _\<rbrace>
+  update_time_stamp
+  \<lbrace>\<lambda>_. budget_ready t\<rbrace>"
+   apply (rule_tac Q="\<lambda>_ s. \<exists>scp. bound_sc_tcb_at (\<lambda>ko. ko = Some scp) t s \<and>
+                                 is_refill_ready scp s"
+                in hoare_strengthen_post)
+   apply (wpsimp wp: hoare_vcg_ex_lift hoare_vcg_imp_lift hoare_vcg_conj_lift)
+   apply (clarsimp simp: vs_all_heap_simps tcb_at_kh_simps)
+  apply (clarsimp simp: vs_all_heap_simps tcb_at_kh_simps)
+  done
+
+lemma update_time_stamp_valid_sched[wp]:
+  "\<lbrace>valid_sched and valid_machine_time\<rbrace>
+   update_time_stamp
+   \<lbrace>\<lambda>_. valid_sched :: 'state_ext state \<Rightarrow> _\<rbrace>"
+  apply (wpsimp wp: valid_sched_lift_pre_conj[where R=valid_machine_time] simp: valid_sched_def)
+  done
+
+lemma preemption_point_valid_sched[wp]:
+  "\<lbrace>valid_sched and valid_machine_time\<rbrace>
+   preemption_point
+   \<lbrace>\<lambda>_. valid_sched :: 'state_ext state \<Rightarrow> _\<rbrace>"
+  apply (clarsimp simp: preemption_point_def)
+  apply (wpsimp wp: OR_choiceE_weak_wp hoare_vcg_all_lift | wp (once) hoare_drop_imps)+
+  done
+
+lemma preemption_point_simple_sched_action[wp]:
+  "preemption_point \<lbrace>simple_sched_action :: 'state_ext state \<Rightarrow> _\<rbrace>"
+  apply (wpsimp simp: preemption_point_def
+                  wp: OR_choiceE_weak_wp hoare_drop_imps)
+  done
+
+crunches finalise_cap
+  for vmt[wp]: "(\<lambda>s. P (last_machine_time_of s) (cur_time s)) :: 'state_ext state \<Rightarrow> _"
```

Comment:
It seems that the acronym `vmt` is used to mean several different things. I am not sure if the variation is introduced in this PR or existing. Ideally we should clean this up, but maybe it is outside of the scope of this PR. We could do with a JIRA issue, maybe.

Reply:
I'm fairly sure `vmt` was used elsewhere. 

I think crunches will actually look at the name when finding rules? I've sometimes seen it complain that it doesn't match existing rules or something. @corlewis would know. So I think the idea here is that we might use `vmt` when crunching the predicate that appears here, which is different from the predicate `valid_machine_time` itself (which I'm guessing is what's used when the name `valid_machine_time` is used)

Reply:
Oh, I guess that didn't answer your question. You probably already knew that. 

Yeah I can make a JIRA issue for it

Reply:
https://sel4.atlassian.net/browse/VER-1393